ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -714,18 +730,21 @@ void checkAndMaybeComplete() {
 
     /**
      * Handle the body of the response: Deserialize and add to the list of chunk buffers.
+     * @param responseInfo the response received for a request sent out on behalf of this chunk.
      * @param payload the body of the response.
      * @param messageMetadata the {@link MessageMetadata} associated with the message.
      * @param messageInfo the {@link MessageInfo} associated with the message.
      * @throws IOException if there is an IOException while deserializing the body.
      * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
      */
-    void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
-        throws IOException, MessageFormatException {
+    void handleBody(ResponseInfo responseInfo, InputStream payload, MessageMetadata messageMetadata,
+        MessageInfo messageInfo) throws IOException, MessageFormatException {
       if (!successfullyDeserialized) {
         BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
         ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
         ByteBuffer chunkBuffer = blobData.getStream().getByteBuffer();
+        responseInfo.retain();
+        chunkIndexToResponseInfo.put(chunkIndex, responseInfo);
 
         boolean launchedJob = maybeLaunchCryptoJob(chunkBuffer, null, encryptionKey, chunkBlobId);
         if (!launchedJob) {
```

Comment:
It will be released in GetManager.

In GetManger, it calls GetBlobOperation.handleResponse and passes the ResponseInfo as the first parameter. After it finishes handleResponse method, it will release ResponseInfo. So if the responseInfo is not retained inside handleResponse function, then we don't need extra release call.